docs/building-modules/module-interfaces.md

```diff
@@ -0,0 +1,93 @@
+  + If applicable, the `CLIContext` is used to retrieve any parameters (e.g. the query originator's address to be used in the query) and marshal them with the query parameter type, in preparation to be relayed to a node.
+  + Use the `queryRoute` to construct a route Baseapp will use to route the query to the appropriate [querier](./querier.md). Module queries are `custom` type queries.  
+  + Call the `CLIContext` query function to relay the query to a node and retrieve the response.
+  + Unmarshal the response and use the `CLIContext` to print the output back to the user.
+- **Flags.** Add any [flags](#flags) to the command.
+
+
+Finally, the module also needs a `GetQueryCmd`, which aggregates all of the query commands of the module. Application developers wishing to include the module's queries will call this function to add them as subcommands in their CLI.
+
+### Flags
+
+[Flags](../interfaces/cli.md#flags) are entered by the user and allow for command customizations. Examples include the fees or gas prices users are willing to pay for their transactions.
+
+The flags for a module are typically found in the `flags.go` file in the `/client/cli` folder. Module developers can create a list of possible flags including the value type, default value, and a description displayed if the user uses a `help` command. In each transaction getter function, they can add flags to the commands and, optionally, mark flags as _required_ so that an error is thrown if the user does not provide values for them.
+
+For full details on flags, visit the [Cobra Documentation](https://github.com/spf13/cobra).
+
+## REST
+
+Application users may find the most intuitive methods of interfacing with the application are web services that use HTTP requests (e.g. a web wallet like [Lunie.io](lunie.io)). Thus, application developers will also use REST Routes to route HTTP requests to the application's modules. The module developer's responsibility is to define the REST client by defining routes for all possible requests and handlers for each of them. The REST interface file is typically found in the module's `/client/rest` folder.
+
+### Request Types
+
+Request types must be defined for all *transaction* requests. Conventionally, each request is named with the suffix `Req`, e.g. `SendReq` for a Send transaction. Each struct should include a base request [`baseReq`](../interfaces/rest.md#basereq), the name of the transaction, and all the arguments the user must provide for the transaction.
+
+### Request Handlers
+
+Request handlers must be defined for both transaction and query requests. Handlers' arguments include a reference to the application's `codec` and the [`CLIContext`](../interfaces/query-lifecycle.md#clicontext) created in the user interaction.
+
+### Register Routes
+
+The entrypoint `RegisterRoutes` function of the application will call the  `registerRoutes` functions of each module utilized by the application.
```

Comment:
what is "the entrypoint `RegisterRoutes`"

-> instead, say precisely where the global `RegisterRoutes` function is called (in which file) + link to it

Reply:
Also explain how to implement the `registerRoutes` function at module level